### PR TITLE
Disable NODEJS_CATCH_EXIT by default. NFC

### PR DESCRIPTION
**Pull request by sbc100**
It turns out that this setting is only needed in very specific
circumstances.  Specifically its only needed if native code calls
`exit` a callback that was not wrapped in callUserCallback.

This change removes the need for this setting from all our unit tests
and disables the setting by default.

Fixes: #17228

### CODE REVIEW

**Review by RReverser — APPROVED** [review]
I don't fully know the implications of removing it for those who need it (in the mentioned circumstances), but personally I always had to disable it manually so 👍.

File: .circleci/config.yml [hunk 1]
```diff
@@ -885,8 +880,10 @@ jobs:
       - upload-test-results
 
   test-mac-arm64:
-    executor: mac-arm64
+    executor: python/default
+    resource_class: macos.m1.medium.gen1
     environment:
+      EMSDK_NOTTY: "1"
```
Comment (kripken) [inline .circleci/config.yml:886]: What are these changes?

File: src/library_browser.js [hunk 2]
```diff
@@ -770,13 +770,10 @@ var LibraryBrowser = {
...
   emscripten_async_load_script: (url, onload, onerror) => {
     url = UTF8ToString(url);
-    onload = {{{ makeDynCall('v', 'onload') }}};
-    onerror = {{{ makeDynCall('v', 'onerror') }}};
-
 #if PTHREADS
     if (ENVIRONMENT_IS_PTHREAD) {
       err(`emscripten_async_load_script("${url}") failed, emscripten_async_load_script is currently not available in pthreads!`);
-      return onerror ? onerror() : undefined;
+      return onerror ? {{{ makeDynCall('v', 'onerror') }}}() : undefined;
```
Comment (kripken) [inline src/library_browser.js:776]: Is this change related?

File: test/common.py [hunk 3]
```diff
@@ -1288,10 +1288,7 @@ def clean_js_output(self, output):
     long_lines = []
 
     def cleanup(line):
-      if len(line) > 2048:
-        # Sanity check that this is really the emscripten program/module on
-        # a single line.
-        assert line.startswith('var Module=typeof Module!="undefined"')
+      if len(line) > 2048 and line.startswith('var Module=typeof Module!="undefined"'):
```
Comment (kripken) [inline test/common.py:1291]: What new pattern of size > 2048 is now being accepted here, aside from the program/module on a single line as the old comment used to say?

File: test/test_fs_after_main.c [hunk 4]
```diff
@@ -60,10 +62,10 @@ int main() {
       counter++;
       if (counter < 5) {
         out('js queueing');
-        setTimeout(looper, 1);
+        safeSetTimeout(looper, 1);
       } else {
         out('js finishing');
-        setTimeout(Module['_finish'], 1);
+        safeSetTimeout(Module['_finish'], 1);
```
Comment (kripken) [inline test/test_fs_after_main.c:68]: Can you explain why this change is needed? A comment might be good in the code too.

File: test/common.py (hunk 3 above)
Comment (sbc100) [reply test/common.py:1291]: We had one test where we have line over 2kb but it was coming from a long line which was the minified wasm2js output, not the overall emscripten output, so it didn't start with `var Module..`

So now, such tests would not get this "REPLACED ENTIRE PROGRAM ON SINGLE LINE" hack.. only when we is clearly the entire emscripten program minified do we do this trick?  We could come up with a pattern to detect wasm2js output but I don't think its worth it.

Comment (sbc100) [reply test/common.py:1291]: BTW the `len(line) > 2048` condition was always there, I just most the assertion up into a condition becuase it was firing.

File: src/library_browser.js (hunk 2 above)
Comment (sbc100) [reply src/library_browser.js:776]: Yes, this is part of the change.  It makes it possible to safely call `exit` from one of the `emscripten_async_load_script` callbacks by wrapping them in `callUserCallback` (rather then directly calling the).

File: .circleci/config.yml (hunk 1 above)
Comment (sbc100) [reply .circleci/config.yml:886]: This fix landed as #22260.  Reverted here.

File: src/library_browser.js (hunk 2 above)
Comment (kripken) [reply src/library_browser.js:776]: How does this make that safe? This change looks NFC to me, like a cleanup - `onerror` is just moved around. What am I missing? Or is this part NFC but the code later is what makes a difference?

File: test/common.py (hunk 3 above)
Comment (kripken) [reply test/common.py:1291]: Yes, I understand that the 2048 is pre-existing, but sorry, I'm still not following what changed here. If the long content is minified wasm2js output, why is it printed out after this PR, but not before?

File: src/library_browser.js (hunk 2 above)
Comment (sbc100) [reply src/library_browser.js:776]: When the callbacks are called asyncronously below they are now called via callUserCallback.  Before they were not.

File: test/common.py (hunk 3 above)
Comment (sbc100) [reply test/common.py:1291]: Because one of the consequences of this change is that when node prints the line that threw an error is not prints a more accurate line, rather then printing the location of the rethrow.

Because we are no longer rethrowing at least on error is now (correctly) reported as coming from inside of wasm2js content rather than the minified emscripten-generated JS.

This code here in common.py avoids echoing the entire emscripten-genertaed module, but it doesn't (currently avoid) echoing the asmjs code (which appears on a different line).

Comment (kripken) [reply test/common.py:1291]: I see now, thanks.

File: ChangeLog.md [hunk 5]
```diff
@@ -20,6 +20,9 @@ See docs/process.md for more on how version tagging works.
 
 3.1.64 (in development)
 -----------------------
+- The `NODEJS_CATCH_EXIT` setting is now disabled by default.  This setting
+  is only useful under very specific circumstances, and has some downsides, so
+  disabling it by default makes sense. (#22257)
```
Comment (kripken) [inline ChangeLog.md:25]:
```suggestion
- The `NODEJS_CATCH_EXIT` setting is now disabled by default.  This setting
  is only useful under very specific circumstances, and has some downsides, so
  disabling it by default makes sense. If you run code in Node.js and want the
  old behavior, pass `-sNODEJS_CATCH_EXIT`. (#22257)
```

**Review by kripken — APPROVED** [review]
